### PR TITLE
Converting the AuthSecret field to a union AuthInfo type

### PR DESCRIPTION
**Pull request by arschles**
This change will enable operators to chose different auth schemes for brokers, as the OSB API spec begins to support more in the future.

Fixes https://github.com/kubernetes-incubator/service-catalog/issues/864

### CODE REVIEW

File: pkg/apis/servicecatalog/types.go [hunk 1]
```diff
@@ -48,9 +48,21 @@ type BrokerSpec struct {
...
+// BrokerAuthInfo is a union type that contains information on one of the authentication methods
+// the the service catalog and brokers may support, according to the OpenServiceBroker API
+// specification (https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md).
+//
+// Note that we currently restrict a single broker to have only one of these fields
+// set on it.
+type BrokerAuthInfo struct {
+	// BasicAuthSecret is a reference to a Secret containing auth information the
+	// catalog should use to authenticate to this Broker using basic auth.
+	BasicAuthSecret *v1.ObjectReference
```
Comment (duglin) [inline pkg/apis/servicecatalog/types.go]: Is the assumption that we know which one is being used because all others will be nil ?
If so, does that require a loop (or set of if-statements) to find the right one?  Could we instead just make `AuthInfo` in the previous struct a pointer to an Object and dynamically find its type at runtime?

Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: @duglin 

> Is the assumption that we know which one is being used because all others will be nil ?

Yes, the generic assertion is that if you provide a union type, you must provide one and only one field of that type.  Since we only have basic auth right now, that collapses into 'you must provide basic auth'

File: pkg/apis/servicecatalog/validation/broker.go [hunk 2]
```diff
@@ -49,13 +49,16 @@ func validateBrokerSpec(spec *sc.BrokerSpec, fldPath *field.Path) field.ErrorLis
 				"brokers must have a remote url to contact"))
 	}
 
-	if spec.AuthSecret != nil {
-		for _, msg := range apivalidation.ValidateNamespaceName(spec.AuthSecret.Namespace, false /* prefix */) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("authSecret", "namespace"), spec.AuthSecret.Namespace, msg))
+	// TODO: when we start supporting additional auth schemes, this code will have to accommodate
+	// the new schemes
+	basicAuthSecret := spec.AuthInfo.BasicAuthSecret
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: This is not nil-safe

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 3]
```diff
@@ -51,9 +51,11 @@ func TestValidateBroker(t *testing.T) {
 				},
 				Spec: servicecatalog.BrokerSpec{
 					URL: "http://example.com",
-					AuthSecret: &v1.ObjectReference{
-						Namespace: "test-ns",
-						Name:      "test-secret",
+					AuthInfo: servicecatalog.BrokerAuthInfo{
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker_test.go]: Needs a test case for AuthInfo not nil but empty

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -1670,11 +1670,15 @@ func (c *controller) getServiceClassPlanAndBrokerForBinding(instance *v1alpha1.I
 // returns an error. If the AuthSecret field is nil, empty values are
 // returned.
 func getAuthCredentialsFromBroker(client kubernetes.Interface, broker *v1alpha1.Broker) (username, password string, err error) {
-	if broker.Spec.AuthSecret == nil {
+	// TODO: when we start supporting additional auth schemes, this code will have to accommodate
+	// the new schemes
+	basicAuthSecret := broker.Spec.AuthInfo.BasicAuthSecret
```
Comment (pmorie) [inline pkg/controller/controller.go]: also not nil-safe

File: pkg/apis/servicecatalog/types.go (hunk 1 above)
Comment (arschles) [inline pkg/apis/servicecatalog/types.go]: @duglin in addition to what @pmorie said, only one will be set at once (as the comment above says), so only one will be non-nil at once

File: pkg/controller/controller.go (hunk 4 above)
Comment (arschles) [inline pkg/controller/controller.go]: this is nil-safe because `AuthInfo` is a `BrokerAuthInfo`, not a `*BrokerAuthInfo`, and there's a check for a `nil` basic auth secret (which is a `*v1.ObjectReference`)

File: pkg/apis/servicecatalog/validation/broker.go (hunk 2 above)
Comment (arschles) [inline pkg/apis/servicecatalog/validation/broker.go]: good call, I'll add an error if `basicAuthSecret == nil`

File: pkg/apis/servicecatalog/validation/broker_test.go (hunk 3 above)
Comment (arschles) [inline pkg/apis/servicecatalog/validation/broker_test.go]: note that `AuthInfo` cannot be nil because it's not a pointer. `BasicAuthInfo` inside of `AuthInfo` can be nil, though, and that needs a test.

**Comment by arschles** [comment]
@pmorie comments addressed and changes made where appropriate. PTAL

**Comment by vaikas-google** [comment]
Looks like the integration test failure is for realsies:
[Build & Unit/Integration Tests] I0523 21:04:01.468765    3764 round_trippers.go:398] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: integration.test/v0.0.0 (linux/amd64) kubernetes/$Format" http://localhost:23533/apis/servicecatalog.k8s.io/v1alpha1/serviceclasses?resourceVersion=0
[Build & Unit/Integration Tests] I0523 21:04:01.469412    3764 request.go:991] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Broker.servicecatalog.k8s.io \"test-broker\" is invalid: spec.authInfo.basicAuthSecret: Required value: a basic auth secret is required","reason":"Invalid","details":{"name":"test-broker","group":"servicecatalog.k8s.io","kind":"Broker","causes":[{"reason":"FieldValueRequired","message":"Required value: a basic auth secret is required","field":"spec.authInfo.basicAuthSecret"}]},"code":422}
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlows (5.07s)
[Build & Unit/Integration Tests] 	controller_test.go:267: controller start
[Build & Unit/Integration Tests] 	controller_test.go:275: informers start
[Build & Unit/Integration Tests] 	controller_test.go:104: error creating the broker &{{"" ""} {"test-broker" "" "" "" "" "" '\x00' "0001-01-01 00:00:00 +0000 UTC" <nil> %!q(*int64=<nil>) map[] map[] [] [] ""} {"https://example.com" {"nil"}} {[]}} ("Broker.servicecatalog.k8s.io \"test-broker\" is invalid: spec.authInfo.basicAuthSecret: Required value: a basic auth secret is required")
[Build & Unit/Integration Tests] 	framework.go:61: Shutting down server on ports: 23533 and 16738
[Build & Unit/Integration Tests] FAIL
[Build & Unit/Integration Tests] exit status 1
[Build & Unit/Integration Tests] FAIL	github.com/kubernetes-incubator/s

File: pkg/apis/servicecatalog/types.go [hunk 5]
```diff
@@ -48,9 +48,21 @@ type BrokerSpec struct {
 	// URL is the address used to communicate with the Broker.
 	URL string
 
-	// AuthSecret is a reference to a Secret containing auth information the
-	// catalog should use to authenticate to this Broker.
-	AuthSecret *v1.ObjectReference
+	// AuthInfo contains the data that the service catalog should use to authenticate
+	// with the Broker
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: missing period

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 6]
```diff
@@ -49,9 +49,21 @@ type BrokerSpec struct {
 	// URL is the address used to communicate with the Broker.
 	URL string `json:"url"`
 
-	// AuthSecret is a reference to a Secret containing auth information the
-	// catalog should use to authenticate to this Broker.
-	AuthSecret *v1.ObjectReference `json:"authSecret,omitempty"`
+	// AuthInfo contains the data that the service catalog should use to authenticate
+	// with the Broker
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: missing period

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 7]
```diff
@@ -32,6 +32,8 @@ func TestValidateBroker(t *testing.T) {
 		valid  bool
 	}{
 		{
+			// covers the case where there is no AuthInfo field specified. the validator should
+			// ignore the field and still succeed the validation
 			name: "valid broker - no auth secret",
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker_test.go]: not a feedback on your PR, but I have had a lot of luck with a method that returns a valid resource, and then subtracting / mutating things to create objects for invalid scenarios.

**Comment by pmorie** [comment]
LGTM
